### PR TITLE
vz 9481 default jaeger install

### PR DESCRIPTION
**Pull request by AbeHern**

### CODE REVIEW

File: platform-operator/controllers/verrazzano/component/jaeger/operator/jaeger_operator_component.go [hunk 1]
```diff
@@ -88,12 +89,16 @@ func NewComponent() spi.Component {
 			SupportsOperatorUninstall: true,
 			MinVerrazzanoVersion:      constants.VerrazzanoVersion1_3_0,
 			ImagePullSecretKeyname:    "image.imagePullSecrets[0]",
-			ValuesFile:                filepath.Join(config.GetHelmOverridesDir(), "jaeger-operator-values.yaml"),
-			Dependencies:              []string{networkpolicies.ComponentName, cmconstants.CertManagerComponentName, opensearch.ComponentName, fluentoperator.ComponentName},
-			AppendOverridesFunc:       AppendOverrides,
-			GetInstallOverridesFunc:   GetOverrides,
+			//ValuesFile:                filepath.Join(config.GetHelmOverridesDir(), "jaeger-operator-values.yaml"),
```
Comment (anders-swanson) [inline platform-operator/controllers/verrazzano/component/jaeger/operator/jaeger_operator_component.go:92]: remove commented line

File: platform-operator/controllers/verrazzano/component/jaeger/operator/jaeger_operator_component.go [hunk 2]
```diff
@@ -77,6 +77,7 @@ var (
 )
 
 func NewComponent() spi.Component {
+	//j := jaegerOperatorComponent{
```
Comment (anders-swanson) [inline platform-operator/controllers/verrazzano/component/jaeger/operator/jaeger_operator_component.go:80]: remove commented line

File: platform-operator/controllers/verrazzano/component/jaeger/operator/jaeger_operator_component.go [hunk 3]
```diff
@@ -88,12 +89,16 @@ func NewComponent() spi.Component {
...
-			Dependencies:              []string{networkpolicies.ComponentName, cmconstants.CertManagerComponentName, opensearch.ComponentName, fluentoperator.ComponentName},
-			AppendOverridesFunc:       AppendOverrides,
-			GetInstallOverridesFunc:   GetOverrides,
+			//ValuesFile:                filepath.Join(config.GetHelmOverridesDir(), "jaeger-operator-values.yaml"),
+			Dependencies:            []string{networkpolicies.ComponentName, cmconstants.CertManagerComponentName, opensearch.ComponentName, fluentoperator.ComponentName},
+			AppendOverridesFunc:     AppendOverrides,
+			GetInstallOverridesFunc: GetOverrides,
 		},
 	}
+	//j.AppendOverridesFunc = func(context spi.ComponentContext, releaseName string, namespace string, chartDir string, kvs []bom.KeyValue) ([]bom.KeyValue, error) {
```
Comment (anders-swanson) [inline platform-operator/controllers/verrazzano/component/jaeger/operator/jaeger_operator_component.go:98]: remove commented lines

File: platform-operator/controllers/verrazzano/component/jaeger/operator/jaeger_operator.go [hunk 4]
```diff
@@ -268,6 +282,21 @@ func AppendOverrides(compContext spi.ComponentContext, _ string, _ string, _ str
 			if err != nil {
 				return nil, err
 			}
+		} else {
+			allInOneImage := filepath.Join(config.GetHelmOverridesDir(), "jaeger-all-in-one-operator-values.yaml")
```
Comment (anders-swanson) [inline platform-operator/controllers/verrazzano/component/jaeger/operator/jaeger_operator.go:286]: values file, not an image?

File: platform-operator/controllers/verrazzano/component/jaeger/operator/jaeger_operator.go [hunk 5]
```diff
@@ -268,6 +287,23 @@ func AppendOverrides(compContext spi.ComponentContext, _ string, _ string, _ str
...
+			createInstance, err = isCreateAllInOneJaegerInstance(compContext)
+			if err != nil {
+				return nil, err
+			}
+			if createInstance {
+				jaegerCRTemplate, err := template.New("jaeger").Parse(jaegerAllInOneTemplate)
+				if err != nil {
+					return nil, err
+				}
+				data := jaegerData{OpenSearchURL: openSearchURL, SecretName: globalconst.DefaultJaegerSecretName, OpenSearchReplicaCount: 0}
```
Comment (prasad-shirodkar) [inline platform-operator/controllers/verrazzano/component/jaeger/operator/jaeger_operator.go:300]: `jaegerAllInOneTemplate ` does not have these 3 fields. Why are you trying to fill them in the template?

File: platform-operator/helm_config/overrides/jaeger-all-in-one-operator-values.yaml [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+      containerSecurityContext:
+        runAsUser: 1001
+        runAsGroup: 65534
+        runAsNonRoot: true
+        privileged: false
+        allowPrivilegeEscalation: false
+        capabilities:
+          drop:
+          - ALL
+    collector:
```
Comment (prasad-shirodkar) [inline platform-operator/helm_config/overrides/jaeger-all-in-one-operator-values.yaml:31]: Jaeger's all-in-one image contains collector and query functionality. Why are you specifying these in the spec?

File: platform-operator/controllers/verrazzano/component/jaeger/operator/jaeger_operator.go [hunk 7]
```diff
@@ -93,6 +94,14 @@ const extraEnvValueTemplate = `extraEnv:
     value: "{{.AllInOneImage}}"
 `
 
+// A template to define Jaeger override for using the JAEGER-ALL-IN-ONE-IMAGE to create a Jaeger instance.
+const jaegerAllInOneTemplate = `jaeger:
+  create: true
+  spec:
+    strategy: allInOne
```
Comment (ayaranga) [inline platform-operator/controllers/verrazzano/component/jaeger/operator/jaeger_operator.go:101]: See if you can use storage as "in-memory" so that the overrides file remains same. See https://www.jaegertracing.io/docs/1.8/deployment/#memory for details.

File: platform-operator/helm_config/overrides/jaeger-all-in-one-operator-values.yaml [hunk 8]
```diff
@@ -0,0 +1,86 @@
+# Copyright (c) 2023, Oracle and/or its affiliates.
```
Comment (ayaranga) [inline platform-operator/helm_config/overrides/jaeger-all-in-one-operator-values.yaml:1]: Why do we need a new override file? We can use `jaeger-operator-values.yaml` itself right?

File: platform-operator/controllers/verrazzano/component/jaeger/operator/jaeger_operator.go (hunk 7 above)
Comment (AbeHern) [reply platform-operator/controllers/verrazzano/component/jaeger/operator/jaeger_operator.go:101]: I tried both `in-memory` and `memory` in the storage option and noticed no difference. They both work in starting up the All-in-one instance

File: platform-operator/controllers/verrazzano/component/jaeger/operator/jaeger_operator.go [hunk 9]
```diff
@@ -93,6 +94,14 @@ const extraEnvValueTemplate = `extraEnv:
...
+    storage: in-memory
```
[the remark this answers is not in the document]
Comment (AbeHern) [reply platform-operator/controllers/verrazzano/component/jaeger/operator/jaeger_operator.go:102]: I tried both `in-memory` and `memory` in the storage option and noticed no difference. They both work in starting up the All-in-one instance

File: platform-operator/helm_config/overrides/jaeger-all-in-one-operator-values.yaml (hunk 8 above)
Comment (ayaranga) [reply platform-operator/helm_config/overrides/jaeger-all-in-one-operator-values.yaml:1]: Are you saying that you still need a new override file `jaeger-all-in-one-operator-values.yaml`? If so, what's the exact issue you are facing?

Comment (AbeHern) [reply platform-operator/helm_config/overrides/jaeger-all-in-one-operator-values.yaml:1]: Using the `in-memory` and/or `memory` as the storage type didn't require the `jaeger-all-in-one-operator-values.yaml` file. I was able to use the existing `jaeger-operator-values.yaml` file and still deploy the all-in-one without issues.

File: platform-operator/controllers/verrazzano/component/jaeger/operator/jaeger_operator.go (hunk 5 above)
Comment (AbeHern) [reply platform-operator/controllers/verrazzano/component/jaeger/operator/jaeger_operator.go]: Removed it

File: platform-operator/helm_config/overrides/jaeger-all-in-one-operator-values.yaml (hunk 6 above)
Comment (AbeHern) [reply platform-operator/helm_config/overrides/jaeger-all-in-one-operator-values.yaml]: removed the file entirely, it was not needed

File: platform-operator/controllers/verrazzano/component/jaeger/operator/jaeger_operator_component.go [hunk 10]
```diff
@@ -88,7 +88,6 @@ func NewComponent() spi.Component {
 			SupportsOperatorUninstall: true,
 			MinVerrazzanoVersion:      constants.VerrazzanoVersion1_3_0,
 			ImagePullSecretKeyname:    "image.imagePullSecrets[0]",
-			ValuesFile:                filepath.Join(config.GetHelmOverridesDir(), "jaeger-operator-values.yaml"),
```
Comment (ayaranga) [inline platform-operator/controllers/verrazzano/component/jaeger/operator/jaeger_operator_component.go:91]: You can revert this change as there are no multiple value override files now.

File: platform-operator/controllers/verrazzano/component/jaeger/operator/jaeger_operator.go [hunk 11]
```diff
@@ -268,6 +281,26 @@ func AppendOverrides(compContext spi.ComponentContext, _ string, _ string, _ str
 			if err != nil {
 				return nil, err
 			}
+		} else {
+			createInstance, err = isCreateAllInOneJaegerInstance(compContext)
```
Comment (ayaranga) [inline platform-operator/controllers/verrazzano/component/jaeger/operator/jaeger_operator.go:285]: Let us make this logic simple. We do not need to add a new function ` isCreateAllInOneJaegerInstance`. In isCreateDefaultJaegerInstance(), you can remove the logic of checking for OpenSearch storage and add `if canUseVZOpenSearchStorage(ctx) {` logic here to choose the appropriate template.

File: platform-operator/controllers/verrazzano/component/jaeger/operator/jaeger_operator.go [hunk 12]
```diff
@@ -268,6 +281,26 @@ func AppendOverrides(compContext spi.ComponentContext, _ string, _ string, _ str
...
+			if err != nil {
+				return nil, err
+			}
+			if createInstance {
+				jaegerValuesFile := filepath.Join(config.GetHelmOverridesDir(), "jaeger-operator-values.yaml")
```
Comment (prasad-shirodkar) [inline platform-operator/controllers/verrazzano/component/jaeger/operator/jaeger_operator.go:290]: When using all-in-one strategy, the collector and query components should not be specified in the Jaeger CR spec.  Refer https://www.jaegertracing.io/docs/1.42/operator/#quick-start---deploying-the-allinone-image.

`esIndexCleaner` should also be disabled since the ES index cleaner job does not need to be run. Refer https://github.com/verrazzano/verrazzano/blob/f6e6f2c82427a49cebdc2ae82b4edd55a87cc2fe/platform-operator/helm_config/overrides/jaeger-operator-values.yaml#L70

Comment (AbeHern) [reply platform-operator/controllers/verrazzano/component/jaeger/operator/jaeger_operator.go:290]: That would mean that there would be no values file used then for the all-in-one strategy?

Comment (prasad-shirodkar) [reply platform-operator/controllers/verrazzano/component/jaeger/operator/jaeger_operator.go:290]: The resulting Jaeger CR spec should still have `ingress` component specified. Refer https://github.com/verrazzano/verrazzano/blob/f6e6f2c82427a49cebdc2ae82b4edd55a87cc2fe/platform-operator/helm_config/overrides/jaeger-operator-values.yaml#L14

The helm overrides should still contain these as well:
- https://github.com/verrazzano/verrazzano/blob/f6e6f2c82427a49cebdc2ae82b4edd55a87cc2fe/platform-operator/helm_config/overrides/jaeger-operator-values.yaml#L3-L7
- https://github.com/verrazzano/verrazzano/blob/f6e6f2c82427a49cebdc2ae82b4edd55a87cc2fe/platform-operator/helm_config/overrides/jaeger-operator-values.yaml#L95-L113

Comment (AbeHern) [reply platform-operator/controllers/verrazzano/component/jaeger/operator/jaeger_operator.go:290]: I don't think I'm getting my fundamental question answered. I'm being told on one hand to remove the `jaeger-all-in-one-operator-values.yaml` and use the `jaeger-operator-values.yaml` file for both the default and all-in-one instance. But without `jaeger-all-in-one-operator-values.yaml` where I can specify the values for query and collector, I'm not sure which plan is desired here.

Comment (prasad-shirodkar) [reply platform-operator/controllers/verrazzano/component/jaeger/operator/jaeger_operator.go:290]: FYI I had already mentioned the below around 2 weeks ago. It is essentially irrespective of the implementation https://github.com/verrazzano/verrazzano/pull/6340#discussion_r1257630181

> Jaeger's all-in-one image contains collector and query functionality. Why are you specifying these in the spec?

Comment (ayaranga) [reply platform-operator/controllers/verrazzano/component/jaeger/operator/jaeger_operator.go:290]: > When using all-in-one strategy, the collector and query components should not be specified in the Jaeger CR spec. Refer https://www.jaegertracing.io/docs/1.42/operator/#quick-start---deploying-the-allinone-image.
> 
> `esIndexCleaner` should also be disabled since the ES index cleaner job does not need to be run. Refer
> 
> https://github.com/verrazzano/verrazzano/blob/f6e6f2c82427a49cebdc2ae82b4edd55a87cc2fe/platform-operator/helm_config/overrides/jaeger-operator-values.yaml#L70

I recommend to check what happens if we specify the collector and query components when using all-in-one strategy. Does it ignore or any error seen or it has to specified under different section? If it ignores, then I think we can leave it in the values file as it is.
For  `esIndexCleaner`,  doesn't it ignore if the storage is memory? When `esIndexCleaner` is enabled, if the Jaeger is creating a index cleaner job irrespective of the storage type, then enabling of esIndexCleaner can be moved out of values file and can be added to templates in Go code.

File: platform-operator/controllers/verrazzano/component/jaeger/operator/jaeger_operator.go (hunk 11 above)
Comment (AbeHern) [reply platform-operator/controllers/verrazzano/component/jaeger/operator/jaeger_operator.go]: I made an attempt to simplify some pieces please review again

File: platform-operator/controllers/verrazzano/component/jaeger/operator/jaeger_operator.go [hunk 13]
```diff
@@ -466,8 +485,7 @@ func isJaegerCREnabled(ctx spi.ComponentContext) (bool, error) {
 	if jaegerCreateOverride != nil {
 		return jaegerCreateOverride.(bool), nil
 	}
-	// Jaeger instance would be created if Verrazzano's OpenSearch can be used as storage
-	return canUseVZOpenSearchStorage(ctx), nil
+	return false, nil
```
Comment (ayaranga) [inline platform-operator/controllers/verrazzano/component/jaeger/operator/jaeger_operator.go:488]: @prasad-shirodkar If we return false here, then Jaeger ingresses would not be created as expected right?

**Review by ayaranga — APPROVED** [review]

File: platform-operator/controllers/verrazzano/component/jaeger/operator/jaeger_operator.go [hunk 14]
```diff
@@ -821,3 +855,34 @@ func newScheme() *runtime.Scheme {
...
+	selector, err := metav1.LabelSelectorAsSelector(&metav1.LabelSelector{
+		MatchLabels: map[string]string{
+			k8sInstanceNameLabel: instanceName,
+		},
+	})
+	if err != nil {
+		return nil, err
+	}
+	return &clipkg.ListOptions{
+		Namespace:     ComponentNamespace,
```
Comment (anders-swanson) [inline platform-operator/controllers/verrazzano/component/jaeger/operator/jaeger_operator.go:869]: Is this something the user can change?

Comment (anders-swanson) [reply platform-operator/controllers/verrazzano/component/jaeger/operator/jaeger_operator.go:869]: Same with the labels, could the jaeger cr override these?

**Review by anders-swanson — APPROVED** [review]

File: platform-operator/controllers/verrazzano/component/jaeger/operator/jaeger_operator.go (hunk 14 above)
Comment (ayaranga) [reply platform-operator/controllers/verrazzano/component/jaeger/operator/jaeger_operator.go:869]: We do not provide an option to override the namespace and we have a webhook validation check to disallow the override of Jaeger instance name. See `validateOverrideYamls()` in jaeger_operator.go for details.
